### PR TITLE
[util] create intermediate rasters

### PR DESCRIPTION
**Pull request by rinkk**
Small tool to interpolate n equidistant raster files between two given rasters.

(Originally done for HHS but EN needs this now, too, so it's probably worth including in OGS.)

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.2.2)
2. [x] Tests covering your feature were added?

### CODE REVIEW

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    std::string filename = output_arg.getValue();
+    for (std::size_t i = 0; i < n; ++i)
+    {
+        std::string const basename = BaseLib::dropFileExtension(filename);
+        std::string const ext = BaseLib::getFileExtension(filename);
+
+        GeoLib::RasterHeader h (h1);
+        GeoLib::Raster r(std::move(h), raster[i].begin(), raster[i].end());
```
Comment (endJunction) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: Why not simply make a copy when passing `h1` to Raster ctor?
```suggestion
        GeoLib::Raster r(h1, raster[i].begin(), raster[i].end());
```

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        }
+        it2++;
+    }
+    if (it2 != dem2->end())
+    {
+        ERR("Error: File 1 is shorter than File 2.");
+        return 1;
+    }
+
+    std::string filename = output_arg.getValue();
```
Comment (endJunction) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]:
```suggestion
    std::string const filename = output_arg.getValue();
```

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    std::unique_ptr<GeoLib::Raster> dem1(
+        FileIO::AsciiRasterInterface::readRaster(input1_arg.getValue()));
+    std::unique_ptr<GeoLib::Raster> dem2(
+        FileIO::AsciiRasterInterface::readRaster(input2_arg.getValue()));
+
+    GeoLib::RasterHeader const h1 = dem1->getHeader();
+    GeoLib::RasterHeader const h2 = dem2->getHeader();
+
+    std::size_t errors_found(0);
```
Comment (endJunction) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: a boolean is sufficient; the number of errors in not used anywhere.

Comment (rinkk) [reply Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: :heavy_check_mark:

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp (hunk 2 above)
Comment (rinkk) [reply Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: :heavy_check_mark:

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp (hunk 1 above)
Comment (rinkk) [reply Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: :heavy_check_mark:

File: Applications/Utils/Tests.cmake [hunk 4]
```diff
@@ -248,7 +248,7 @@ MeshTest(
     NAME GocadTSurface_Mesh_Test
     PATH MeshLib/
     EXECUTABLE GocadTSurfaceReader
-    EXECUTABLE_ARGS -i Top-Lower-Shaly.ts -o ${Data_BINARY_DIR}/MeshLib -b
+    EXECUTABLE_ARGS -i Top-Lower-Shaly.ts -o ${Data_BINARY_DIR}/MeshLib -b
```
Comment (endJunction) [inline Applications/Utils/Tests.cmake]: trailing ws;
try not to change unrelated lines.

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        FileIO::AsciiRasterInterface::readRaster(input2_arg.getValue()));
+
+    GeoLib::RasterHeader const h1 = dem1->getHeader();
+    GeoLib::RasterHeader const h2 = dem2->getHeader();
+
+    bool errors_found(false);
+    if (h1.origin[0] != h2.origin[0])
+    {
+        ERR("Origin x-coordinate is not the same in both raster files.\n");
+        errors_found = true;
```
Comment (TomFischer) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: Why not `return EXIT_FAILURE;`?

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            double const step = (max - min) / static_cast<double>(n + 1);
+            for (std::size_t i = 0; i < n; ++i)
+                raster[i].push_back(min + ((i+1) * step));
+        }
+        it2++;
+    }
+    if (it2 != dem2->end())
+    {
+        ERR("Error: File 1 is shorter than File 2.");
+        return 1;
```
Comment (TomFischer) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: `return EXIT_FAILURE;`?

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    for (std::size_t i = 0; i < n; ++i)
+    {
+        std::string const basename = BaseLib::dropFileExtension(filename);
+        std::string const ext = BaseLib::getFileExtension(filename);
+
+        GeoLib::Raster r(h1, raster[i].begin(), raster[i].end());
+        FileIO::AsciiRasterInterface::writeRasterAsASC(r, basename + std::to_string(i) + "." + ext);
+        INFO("Layer %d written.", i+1);
+    }
+    return 0;
```
Comment (TomFischer) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: `return EXIT_SUCCESS;`?

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    cmd.add(number_arg);
+
+    cmd.parse(argc, argv);
+
+    std::unique_ptr<GeoLib::Raster> dem1(
+        FileIO::AsciiRasterInterface::readRaster(input1_arg.getValue()));
+    std::unique_ptr<GeoLib::Raster> dem2(
+        FileIO::AsciiRasterInterface::readRaster(input2_arg.getValue()));
+
+    GeoLib::RasterHeader const h1 = dem1->getHeader();
```
Comment (TomFischer) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: Could `h1` and `h2` be const references?

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp (hunk 5 above)
Comment (rinkk) [reply Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: Different errors give different error codes.

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp (hunk 6 above)
Comment (rinkk) [reply Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: Different errors give different error codes.

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp (hunk 7 above)
Comment (rinkk) [reply Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: :heavy_check_mark:

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp (hunk 8 above)
Comment (rinkk) [reply Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: :heavy_check_mark:

File: Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+#include "GeoLib/Raster.h"
+
+int main(int argc, char* argv[])
+{
+    ApplicationsLib::LogogSetup logog_setup;
+
+    TCLAP::CmdLine cmd(
+        "Takes two DEMs located at the exact same spatial position (but at "
+        "different elevation) and calculates n raster DEMs located at "
+        "equidistant intervalls between them (i.e. for n=1, one new raster "
```
Comment (endJunction) [inline Applications/Utils/MeshGeoTools/createIntermediateRasters.cpp]: intervals

**Comment by rinkk** [comment]
did the rebase but something obvs went wrong. no idea what happened. pls halp?!

**Comment by TomFischer** [comment]
I rebased and fixed the spelling. I'll merge if all tests are green.
